File: swift/Sources/DuetShells/PresentationRegistry.swift
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

import SwiftUI

// W1 — Layer 3 of the presentation contract (parity/presentation-contract.md §5): the
// per-HOST kind→renderer registry. A feature ships its kind (an enum case, Layer 1);
// the platform's composition root binds how that kind renders — manner (detents,
// chrome, wrapping controllers) lives entirely inside the renderer closure, so
// reducers and fixtures never see it. One registry per interior host node; a node
// whose slot has a single kind may render directly (a one-entry registry is ceremony,
// not a seam) — the registry earns its keep where kinds accumulate.

/// Kind-indexed renderer table. Registration is composition-root work; resolution is
/// host-view work. Renderers may return nil (child not built yet — hosts fall back to
/// a placeholder and the reconciler catches up).
@MainActor
public final class PresentationRegistry {
  private var renderers: [ObjectIdentifier: (AnyHashable) -> AnyView?] = [:]

  public init() {}

  /// Bind the renderer for a kind type. Re-registering replaces (last write wins —
  /// composition roots register once at construction).
  public func register<Kind: Hashable>(
    _ kind: Kind.Type, renderer: @escaping (Kind) -> AnyView?
  ) {
    renderers[ObjectIdentifier(kind)] = { anyKind in
      guard let kind = anyKind.base as? Kind else { return nil }
      return renderer(kind)
    }
  }

  /// Resolve the platform surface for a kind value; nil when the kind type has no
  /// renderer or the renderer declined (child not reconciled yet).
  public func view<Kind: Hashable>(for kind: Kind) -> AnyView? {
    renderers[ObjectIdentifier(Kind.self)]?(AnyHashable(kind))
  }
}

File: kotlin/shells-compose/src/main/kotlin/dev/modaal/duet/shells/PresentationRegistry.kt
// Copyright (c) 2026 Modaal.dev
// Licensed under the MIT License. See LICENSE file for details.

package dev.modaal.duet.shells

import kotlin.reflect.KClass

// Layer 3 of the presentation contract (contracts/presentation-contract.md §5):
// the per-HOST kind→renderer registry. A feature ships its kind (a sealed case,
// Layer 1); the platform's composition root binds how that kind renders —
// manner (detents, chrome, wrapping surfaces) lives entirely inside the
// renderer closure, so reducers and fixtures never see it. One registry per
// interior host node; a node whose slot has a single kind may render directly
// (a one-entry registry is ceremony, not a seam).

/**
 * Kind-indexed renderer table. Registration is composition-root work;
 * resolution is host-view work. Renderers may return null (child not built yet
 * — hosts fall back to a placeholder and the reconciler catches up).
 *
 * Thinness note (vs the Swift twin): generic over [Surface] where Swift pins
 * `AnyView` — SwiftUI forces an erasure type; Kotlin needs none, and staying
 * generic keeps this artifact Compose-free (a Compose app instantiates
 * `PresentationRegistry<@Composable () -> Unit>` — or its own surface value —
 * from its composition roots). Swift-flavor mirror:
 * DuetShells/PresentationRegistry.swift.
 */
class PresentationRegistry<Surface : Any> {
  private val renderers = mutableMapOf<KClass<*>, (Any) -> Surface?>()

  /**
   * Bind the renderer for a kind type. Re-registering replaces (last write
   * wins — composition roots register once at construction).
   */
  fun <Kind : Any> register(kind: KClass<Kind>, renderer: (Kind) -> Surface?) {
    renderers[kind] = { anyKind ->
      @Suppress("UNCHECKED_CAST")
      (anyKind as? Kind)?.let(renderer)
    }
  }

  /**
   * Resolve the platform surface for a kind value; null when the kind type has
   * no renderer or the renderer declined (child not reconciled yet).
   *
   * Sealed-hierarchy note (the Swift twin resolves by STATIC type; Kotlin
   * values carry their leaf class): an exact-class entry wins, else the first
   * registered entry whose kind class the value is an instance of — so
   * registering the sealed ROOT covers every case, matching the Swift
   * geometry.
   */
  fun surfaceFor(kind: Any): Surface? {
    val entry =
      renderers[kind::class]
        ?: renderers.entries.firstOrNull { it.key.isInstance(kind) }?.value
    return entry?.invoke(kind)
  }
}
